src/common/utils.js

```diff
@@ -583,3 +592,69 @@ export const createFakeProgress = (() => {
     }
   }
 })()
+
+// ===================================================================
+
+// Smart backup
+export const destructPattern = (pattern, valueTransform = identity) => pattern && ({
+  not: !!pattern.__not,
+  values: valueTransform((pattern.__not || pattern).__or)
+})
+
+export const constructPattern = ({ not, values } = EMPTY_OBJECT, valueTransform = identity) => {
+  if (values == null || !values.length) {
+    return
+  }
+
+  const pattern = { __or: valueTransform(values) }
+  return not
+    ? { __not: pattern }
+    : pattern
+}
+
+const parsePattern = pattern => {
+  const patternValues = map(pattern.values, value => {
+    return isArray(value)
+      ? createString(value[0])
+      : createString(value)
+  })
+
+  return pattern.not
+    ? createNot(createOr(patternValues))
+    : createOr(patternValues)
+}
+
+export const constructFilter = pattern => {
+  const powerState = pattern.power_state
+  const pool = destructPattern(pattern.$pool)
+  const tags = destructPattern(pattern.tags)
+
+  let filter = []
+
+  if (powerState !== undefined) {
+    filter.push(
+      createProperty(
+        'power_state',
+        powerState === 'Running' ? createString('running') : createNot(createString('running'))
+      )
+    )
+  }
+  if (pool !== undefined) {
+    filter.push(
+      createProperty(
+        '$pool',
+        parsePattern(pool)
+      )
+    )
+  }
+  if (tags !== undefined) {
+    filter.push(
+      createProperty(
+        'tags',
+        parsePattern(tags)
+      )
+    )
+  }
+
+  return createAnd(filter)::toString()
+}
```

Comment:
This code needs to be extracted from xo-server/xo-web into xo-common.

Reply:
This code is not used on the server side.

Reply:
There going to be some changes: the complex matcher has been extracted from xo-web and it's API has changed a bit in #2536.